src/operators/batch.rs

```diff
@@ -0,0 +1,75 @@
+use std::{
+    task::Poll,
+    time::{Duration, Instant},
+};
+
+use pyo3::{Python, ToPyObject};
+
+use crate::{pyo3_extensions::TdPyAny, unwrap_any};
+
+use super::stateful_unary::{LogicFate, StatefulLogic};
+
+pub(crate) struct BatchLogic {
+    size: usize,
+    timeout: Duration,
+    start: Instant,
+    acc: Vec<TdPyAny>,
+}
+
+impl BatchLogic {
+    pub(crate) fn builder(size: usize, timeout: Duration) -> impl Fn(Option<TdPyAny>) -> Self {
+        move |resume_snapshot| {
+            let acc = resume_snapshot
+                .and_then(|state| -> Option<Vec<TdPyAny>> {
+                    unwrap_any!(Python::with_gil(|py| state.extract(py)))
+                })
+                // Since we know the max size of the vec, allocate
+                // with all the necessary capacity here.
+                .unwrap_or_else(|| Vec::with_capacity(size));
```

Comment:
uh, didn't think about that, back to usual `Vec::new` then. Also, this makes me think that maybe we should allow to set only one of the two limits, so that users won't have to put an arbitrary big size of the batch if they only want to batch by time (or the inverse).